File: client/lib/data/services/post_service.dart
import 'package:dio/dio.dart';
import '../../core/network/api_client.dart';

class PostService {
  // Lấy danh sách bài đăng
  Future<Response> getPosts({int page = 1, int limit = 10}) async {
    return await ApiClient.dio.get('/posts', queryParameters: {
      'page': page,
      'limit': limit,
    });
  }

  // Tạo bài đăng mới (có thể kèm ảnh)
  Future<Response> createPost({
    required String type,
    required String content,
    String? sport,
    String? location,
    String? time,     // ISO 8601 string, VD: "2026-05-10T19:00:00.000Z"
    String? imagePath,
  }) async {
    // Dùng FormData vì có thể có upload ảnh
    final formData = FormData.fromMap({
      'type': type,
      'content': content,
      if (sport != null) 'sport': sport,
      if (location != null) 'location': location,
      if (time != null) 'time': time,
      if (imagePath != null)
        'image': await MultipartFile.fromFile(imagePath, filename: 'post_image.jpg'),
    });
    return await ApiClient.dio.post('/posts', data: formData);
  }

  // Like / Bỏ Like
  Future<Response> likePost(String postId) async {
    return await ApiClient.dio.put('/posts/$postId/like');
  }

  // Thêm comment
  Future<Response> addComment(String postId, String text) async {
    return await ApiClient.dio.post('/posts/$postId/comments', data: {
      'text': text,
    });
  }
}

File: server/src/services/post.service.js
const postRepository = require('../repositories/post.repository');

class PostService {

    // Tạo bài đăng mới
    async createPost(userId, body, imageUrl) {
        const { type, content, sport, location, time } = body;

        // Validate: Nếu là bài tìm kèo thì phải đủ thông tin
        if (type === 'MATCH') {
            if (!sport || !location || !time) {
                const error = new Error('Bài tìm kèo cần có môn thể thao, địa điểm và thời gian');
                error.statusCode = 400;
                throw error;
            }
        }

        const postData = {
            author: userId,
            type: type || 'DISCUSSION',
            content,
            image: imageUrl || '',
            sport: sport || undefined,
            location: location || undefined,
            time: time ? new Date(time) : undefined,
        };

        return await postRepository.createPost(postData);
    }

    // Lấy danh sách bài đăng (có phân trang)
    async getPosts(page = 1, limit = 10) {
        const skip = (page - 1) * limit;
        return await postRepository.getPosts(skip, limit);
    }

    // Like / Bỏ Like
    async likePost(postId, userId) {
        return await postRepository.toggleLike(postId, userId);
    }

    async getPostById(postId) {
        return await postRepository.getPostById(postId);
    }


    // Thêm comment
    async addComment(postId, userId, text) {
        if (!text || text.trim() === '') {
            const error = new Error('Nội dung comment không được để trống');
            error.statusCode = 400;
            throw error;
        }
        return await postRepository.addComment(postId, userId, text.trim());
    }
}

module.exports = new PostService();
